src/Documentation/implementation/streams_implementation/azure_queue_streams.md

```diff
@@ -53,3 +53,62 @@ hostBuilder
+2. Since streams process messages in order per stream, the gating factor will be the number of events being sent on a single stream.
+3. A reasonable balance of **cache time** (`StreamPullingAgentOptions.GetQueueMsgsTimerPeriod`) vs **visibility time** (`AzureQueueOptions.MessageVisibilityTimeout`) is that the visibility should be configured **to double** the time messages **are expected** to be in the cache.
+
+### Example
+
+We assume a system with this characteristics:
+
+* 100 streams,
+* 10 queues,
+* Each stream processing 60 messages per minute,
+* Each message takes around 30ms to process,
+* 1 minute worth of messages in cache (cache time).
+
+So we can calculate some parameters of the system:
+
+* Streams/queue: Even balancing of streams across queues would be an ideal 10 streams/queue (100 streams / 10 queues).
+But since streams won't always be evenly balanced over the queues, doubling (or even tripling) the ideal is safer than expecting ideal distribution.
+Hence **20 streams/queue** (10 streams/queue x 2 as safety factor) is probably reasonable.
+
+* Messages/minute: This means each queue will be expected to process up to **1200 messages/minute** (60 messages x 20 streams).
+
+The we can determine the visibility time to use:
+
+* Visibility time: The cache time (1 minute) is configured to hold 1 minute of messages (so 1200 messages, as we calculated messages/minute above).
+We assumed that each message takes 30 ms to process, then we can expect messages to spend up to 36 seconds in the cache (0.030 sec x 1200 msg = 36 sec), so the visibility time - doubled for safety - would need be over 72 seconds (36 sec of time in cache x 2).
+Accordingly, if we define a bigger cache, that would require a longer visibility time.
+
+Final considerations in a real world system:
+
+* Since order is only per stream, and a queue consume many streams, it's likely that messages will be processed across multiple streams in parallel (as example: we have a grain for stream, which can run in parallel).
+This means we'll burn through the cache in far less time, but we planned for the worse case: it will give the system room to continue to function well even under intermittent delays and transient errors.
+
+So we can configure Azure Queue Streams using:
+
+``` csharp
+hostBuilder
+  .AddAzureQueueStreams("AzureQueueProvider", configurator => {
+    configurator.ConfigureAzureQueue(
+      ob => ob.Configure(options => {
+        options.ConnectionString = "xxx";
+        options.QueueNames = new List<string> {
+          "yourprefix-azurequeueprovider-1",
+          [...]
+          "yourprefix-azurequeueprovider-10",
+        };
+        options.MessageVisibilityTimeout = TimeSpan.FromSeconds(72);
+      }));
+    configurator.ConfigurePullingAgent(ob => ob.Configure(options => {
+      options.GetQueueMsgsTimerPeriod = TimeSpan.FromMinutes(1);
```

Comment:
Since nothing in this post explains the GetQueueMsgsTimerPeriod, should we remove this setting from the example?  Most of the others should be understood (like ConnectionString) or are relent to the example (like QueueNames and MessageVisibilityTimeout).  Only GetQueueMsgsTimerPeriod seems out of place for this explanation.

Reply:
Then I misunderstood something... It looks like I wrongly supposed that "1 minute worth of messages in cache (cache time)" could be set - looking to the source code - by GetQueueMsgsTimerPeriod. If this is not the case, which setting can we use to configure it?

Reply:
Ah, I see.  Yeah I missed that too.  Good catch.  AzureQueueStreams (or any persistent stream provider using the SimpleQueueCache) configures cache by size, not time, using ConfigureCacheSize.  This can be a bit confusing because persistent stream providers which use the PooledQueueCache are configured by time rather than size, so it's easy to get confused as it's not always clear which cache a persistent stream provider is using.  Originally both were configured by size, but we changed the PooledQueueCache to use time.  Eventually I'd like to get rid of the SimpleQueueCache to avoid the confusion, but I digress.

From the example the cache size for 1 minute of data would be 1200 (default is 4k I think).  So I think we'd need something like:

    configurator.ConfigureCacheSize(1200);

Regarding GetQueueMsgsTimerPeriod: GetQueueMsgsTimerPeriod configures how often we query the queue when no data is returned or the cache is full.  The pulling agent will read from the queue as long as the cache is not full and the queue is returning data.  If the queue is empty, or the cache is full we don't want to free spin checking as that will burn up a lot of cpu, so this setting controls how often we retry.  By default it's 100ms.